spack-ext/lib/jcsda-emc/spack-stack/stack/stack_env.py

```diff
@@ -232,12 +253,22 @@ def write(self):
                 original_sections[key] = copy.deepcopy(section)
 
         # Commonly used config settings
-        if self.compiler:
-            compiler = "packages:all::compiler:[{}]".format(self.compiler)
-            spack.config.add(compiler, scope=env_scope)
-        if self.mpi:
-            mpi = "packages:all::providers:mpi:[{}]".format(self.mpi)
-            spack.config.add(mpi, scope=env_scope)
+        compiler = f"packages:all:require:['%{self.compiler}']"
```

Comment:
This doesn't work. When used in conjuction with, e.g., `packages:gcc-runtime:require:'%gcc'`, this overall setting doesn't get respected. In a test concretization just now, hdf5 and other packages are getting built with %gcc. This is something I've run into previously. The solution appears to be to use `prefer` instead of `require` here (but continue to use `require` for individual packages like gcc-runtime).

Reply:
Actually, I think we can remove the `packages:gcc-runtime:require:'%gcc'` because that constraint is already present in the gcc-runtime recipe itself.

Reply:
This is supposed to work (and it did for me). E4S does the same, use require self.compiler forces using the compiler you want for all packages, but individual packages can overwrite. Which is what we want - we WANT gcc-runtime to be built with gcc always, regardless of what the principal compiler for the eng is.

Reply:
@AlexanderRichert-NOAA See here for the E4S config: https://github.com/spack/spack/blob/develop/share%2Fspack%2Fgitlab%2Fcloud_pipelines%2Fstacks%2Fe4s-oneapi%2Fspack.yaml#L64

Reply:
Steps to reproduce on my CentOS 8 PC (commit 445aca64a55c562db6a90913e075318ff4d692ae):
```console
 $ . setup.sh
 $ spack stack create env --name test-env --template unified-dev --site linux.default --compiler oneapi
 $ spack env activate envs/test-env
 $ spack compiler find # detect oneapi
 $ spack compiler list
==> Available compilers
-- gcc centos8-x86_64 -------------------------------------------
gcc@8.5.0

-- oneapi centos8-x86_64 ----------------------------------------
oneapi@2024.0.0
 $ spack concretize | grep %gcc
 -   shcc26y          ^gcc-runtime@8.5.0%gcc@8.5.0 build_system=generic arch=linux-centos8-zen
 -   xbichoa              ^hdf5@1.14.3%gcc@8.5.0~cxx+fortran+hl~ipo~java~map+mpi+shared~subfiling~szip+threadsafe+tools api=default build_system=cmake build_type=Release generator=make patches=82088c8 arch=linux-centos8-zen
 -   ktbqw7d                  ^cmake@3.27.9%gcc@8.5.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-centos8-zen
 -   r43pbyy          ^bison@3.8.2%gcc@8.5.0~color build_system=autotools arch=linux-centos8-zen
 -   r43pbyy                      ^bison@3.8.2%gcc@8.5.0~color build_system=autotools arch=linux-centos8-zen
...
```

I can also reproduce this problem on Hera. If I add oneapi to my config and concretize, bison, hdf5, and python end up being spec'd with '%intel'. If I take intel out of the available compilers, then those packages end up as %gcc. I think somehow the %oneapi conflict in the bison recipe leads to the general requirement being nullified.

Reply:
Interesting. If you look at the packages in common/packages.yaml that I set to require `%gcc`, then you see that this is bison and everything that bison depends on. hdf5 shouldn't be affected at all, and neither should be cmake. Weird, because I don't have this problem on my Oracle Linux 9. What did you say works for you? I unfortunately don't have my Dell with me on vacation, therefore I won't be able to test it there (but I can test on Ubuntu 22)

Reply:
We should probably report this to spack as a bug as well?

Reply:
I'll retest with 'prefer' and file a bug.

Reply:
I am testing the "prefer" version too. So far looking ok on macOS with apple-cland and on Ubuntu with oneapi (plus the CI tests - just need to verify it used the compilers as intended).